timemachine/cpp/src/random_sampler.cu

```diff
@@ -0,0 +1,89 @@
+#include "device_buffer.hpp"
+#include "gpu_utils.cuh"
+#include "kernels/k_sampling.cuh"
+#include "kernels/kernel_utils.cuh"
+#include "math_utils.cuh"
+#include "random_sampler.hpp"
+
+namespace timemachine {
+
+template <typename RealType>
+RandomSampler<RealType>::RandomSampler(const int N, const int seed)
+    : N_(N), temp_storage_bytes_(0), d_rand_(round_up_even(N_)), d_gumbel_(N_) {
+
+    cudaSafeMalloc(&d_arg_max_, 1 * sizeof(cub::KeyValuePair<int, RealType>));
+    gpuErrchk(cub::DeviceReduce::ArgMax(
+        nullptr, temp_storage_bytes_, d_gumbel_.data, static_cast<cub::KeyValuePair<int, RealType> *>(d_arg_max_), N_));
+    d_sort_storage_.reset(new DeviceBuffer<char>(temp_storage_bytes_));
+
+    // Allocate enough space for a single key value pair
+
+    curandErrchk(curandCreateGenerator(&cr_rng_, CURAND_RNG_PSEUDO_DEFAULT));
+    curandErrchk(curandSetPseudoRandomGeneratorSeed(cr_rng_, seed));
+};
+
+template <typename RealType> RandomSampler<RealType>::~RandomSampler() {
+    curandErrchk(curandDestroyGenerator(cr_rng_));
+
+    gpuErrchk(cudaFree(d_arg_max_));
+};
+
+template <typename RealType>
+void RandomSampler<RealType>::sample_device(
+    const int N, const int K, const RealType *d_log_probabilities, int *d_samples, cudaStream_t stream) {
+    if (N > N_) {
+        throw std::runtime_error(
+            "N is greater than buffer size:  N=" + std::to_string(N) + ", N_=" + std::to_string(N_));
+    }
+    if (K <= 0) {
+        throw std::runtime_error("K must be at least 1: K=" + std::to_string(K));
+    }
+    // Perform sampling using the gumbel-max-trick like cupy
+    // https://github.com/cupy/cupy/blob/46f6ed6e78661fb1d31b41a072519f119f5a2385/cupy/random/_generator.py#L1115-L1121
+    curandErrchk(curandSetStream(cr_rng_, stream));
+    const int tpb = DEFAULT_THREADS_PER_BLOCK;
+    const int blocks = ceil_divide(N, tpb);
+    for (int i = 0; i < K; i++) {
```

Comment:
N is number of waters, K here is batch size?

Reply:
wish we had a way to specify reduction axis for `cub`

Reply:
Correct. Note that generating multiple samples is not actually done in parallel, just seemed like a simple method to be able to sample multiple indices for testing even if it isn't the most performant.

Reply:
Renamed `K` to `num_samples` to make this clearer